src/Processors/QueryPlan/Optimizations/optimizeReadInOrder.cpp

```diff
@@ -1172,9 +1200,206 @@ void optimizeAggregationInOrder(QueryPlan::Node & node, QueryPlan::Nodes &)
+static Permutation findCommonOrderInfo(StepInputOrder & left_order_info, StepInputOrder & right_order_info)
+{
+    if (!left_order_info.input_order && !right_order_info.input_order)
+    {
+        LOG_TRACE(getLogger(), "Cannot read anything in order for join");
+        return {};
+    }
+
+    if (!right_order_info.input_order)
+    {
+        LOG_TRACE(getLogger(), "Can read left stream in order for join");
+        return flattenPositions(left_order_info.permutation);
+    }
+
+    if (!left_order_info.input_order)
+    {
+        LOG_TRACE(getLogger(), "Can read right stream in order for join");
+        return flattenPositions(right_order_info.permutation);
+    }
+
+    LOG_TRACE(getLogger(), "Can read both streams in order for join");
+
+    bool left_is_better = compareOrderInfos(left_order_info.input_order, right_order_info.input_order);
+    auto & lhs = left_is_better ? left_order_info : right_order_info;
+    auto & rhs = left_is_better ? right_order_info : left_order_info;
+
+    Permutation result_permutation = flattenPositions(lhs.permutation);
+    size_t prefix_size = findPrefixMatchLength(rhs.permutation, result_permutation);
+    truncateToPrefixSize(rhs, prefix_size);
+    return result_permutation;
+}
+
+static bool optimizeJoinInOrder(QueryPlan::Node & node, const std::shared_ptr<FullSortingMergeJoin> & join_ptr)
+{
+    const auto & key_names_left = join_ptr->getKeyNames(JoinTableSide::Left);
+    auto * left_child_node = node.children[0];
+    QueryPlan::Node * left_reading_node = findReadingStep(*left_child_node->children.front());
+    auto left_order_info = buildInputOrderInfo(key_names_left, *left_child_node, left_reading_node);
+
+    const auto & key_names_right = join_ptr->getKeyNames(JoinTableSide::Right);
+    auto * right_child_node = node.children[1];
+    QueryPlan::Node * right_reading_node = findReadingStep(*right_child_node->children.front());
+    auto right_order_info = buildInputOrderInfo(key_names_right, *right_child_node, right_reading_node);
+
+    auto keys_permuation = findCommonOrderInfo(left_order_info, right_order_info);
+    if (keys_permuation.empty())
+        return false;
+
+    LOG_TRACE(getLogger(), "Applying permutation [{}] for join keys", fmt::join(keys_permuation, ", "));
+
+    join_ptr->permuteKeys(keys_permuation);
+
+    if (left_order_info.input_order)
+    {
+        requestInputOrderInfo(left_order_info.input_order, left_reading_node->step);
+        join_ptr->setPrefixSortDesctiption(left_order_info.sort_description_for_merging, JoinTableSide::Left);
+    }
+
+    if (right_order_info.input_order)
+    {
+        requestInputOrderInfo(right_order_info.input_order, right_reading_node->step);
+        join_ptr->setPrefixSortDesctiption(right_order_info.sort_description_for_merging, JoinTableSide::Right);
+    }
+    return true;
+}
+
+void applyOrderForJoin(QueryPlan::Node & node, QueryPlan::Nodes & nodes, const QueryPlanOptimizationSettings & optimization_settings)
```

Comment:
Please move code related only to `applyOrderForJoin` in separate file. It'd much easier to read it.